src/Microsoft.AspNet.OData.Shared/Builder/InstanceAnnotationDictionaryAnnotation.cs

```diff
@@ -0,0 +1,48 @@
+﻿// Copyright (c) Microsoft Corporation.  All rights reserved.
+// Licensed under the MIT License.  See License.txt in the project root for license information.
+
+using System.Collections.Generic;
+using System.Reflection;
+using Microsoft.AspNet.OData.Common;
+using Microsoft.OData.Edm;
+
+namespace Microsoft.AspNet.OData.Builder
+{
+    /// <summary>
+    /// This annotation indicates the mapping from a <see cref="IEdmStructuredType"/> to a <see cref="PropertyInfo"/>.
+    /// The <see cref="IEdmStructuredType"/> is an open type and the <see cref="PropertyInfo"/> is the specific
+    /// property which is used in an open type to save/retrieve the instance annotations.
+    /// </summary>
+    public class InstanceAnnotationDictionaryAnnotation
+    {
+        /// <summary>
+        /// Initializes a new instance of <see cref="InstanceAnnotationDictionaryAnnotation"/> class.
+        /// </summary>
+        /// <param name="propertyInfo">The backing <see cref="PropertyInfo"/>.</param>
+        public InstanceAnnotationDictionaryAnnotation(PropertyInfo propertyInfo)
+        {
+            if (propertyInfo == null)
+            {
+                throw Error.ArgumentNull("propertyInfo");
+            }
+
+            if (!typeof(IDictionary<string,IDictionary<string, object> >).IsAssignableFrom(propertyInfo.PropertyType))
```

Comment:
>IDictionary<string,IDictionary<string, object> > [](start = 24, length = 48)

What i am thinking is that: can we have an interface design.

pubic interface IODataInstanceAnnotationContainer
{
         void AddAnnotation(string annotationName, object value, bool forType);
         object GetAnnotation(string annotationName, bool forType);
}

Then, developer can use this interface to define the container property in the CLR type?

Reply:
Then they should be able to have their own implementation right? But I guess thats not intended here? right?

Reply:
I like this idea.  We could even have separate methods for getting/adding annotations to the object versus a property:

pubic interface IODataInstanceAnnotationContainer
{
         void AddAnnotation(string annotationName, object value);
         void AddPropertyAnnotation(string propertyName, string annotationName, object value);
         object GetAnnotation(string annotationName, bool forType);
         object GetPropertyAnnotation(string propertyName, string annotationName, bool forType);
}

And provide a default implementation ODataInstanceAnnotationContainer the exposes this on top of an IDictionary<string,IDictionary<string,object>>

---
In reply to: [470879810](https://github.com/OData/WebApi/pull/2219#discussion_r470879810) [](ancestors = 470879810)